frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py

```diff
@@ -0,0 +1,851 @@
+    # We must change it to sample with a single shot.
+
+    # Otherwise, we will be returning a quake value
+    # that is opaque and cannot be inspected for a value by the user.
+    # For the time being, we can just add an exception if the return of
+    # measurement is being returned directly.
+    ctx.add_measurement(result)
+
+
+def change_expval(ctx, eqn):
+    """Change Catalyst's expval to CUDA-quantum equivalent."""
+    assert eqn.primitive == expval_p
+
+    # Operands to expval_p
+    # * obs: Observables
+    invals = safe_map(ctx.read, eqn.invars)
+    obs = invals[0]
+
+    # Params:
+    # * shots: Shots
+    shots = eqn.params["shots"]
+    shots = shots if shots is not None else -1
+
+    # To obtain expval, we first obtain an observe object.
+    observe_results = cudaq_observe(ctx.kernel, obs, shots)
+    # And then we call expectation on that object.
+    result = cudaq_expectation(observe_results)
+    outvariables = [ctx.new_variable(jax.core.ShapedArray([], float))]
+    outvals = [result]
+
+    safe_map(ctx.replace, eqn.outvars, outvariables)
+    safe_map(ctx.write, eqn.outvars, outvals)
+
+
+def change_namedobs(ctx, eqn):
+    """Change named observable to CUDA-quantum equivalent."""
+    assert eqn.primitive == namedobs_p
+
+    # Operands to expval_p
+    # * qubit
+    invals = safe_map(ctx.read, eqn.invars)
+    qubit = invals[0]
+
+    # Since CUDA doesn't use SSA for qubits.
+    # We now need to get an integer from this qubit.
+    # This will be the target to cudaq_spin....
+
+    # How do we get the wire from the SSA value?
+    # We need to keep a map.
+    # Where for every qubit variable, it maps to an integer.
+    # This is not too hard.
+    idx = ctx.get_wire_for_qubit(qubit)
+
+    # Parameters
+    # * kind
+    kind = eqn.params["kind"]
+    catalyst_cuda_map = {"PauliZ": "z", "PauliX": "x", "PauliY": "y"}
+
+    # This is an assert because it is guaranteed by Catalyst.
+    assert kind in catalyst_cuda_map
+
+    cuda_name = catalyst_cuda_map[kind]
+    outvals = [cudaq_spin(idx, cuda_name)]
+    outvariables = [ctx.new_variable(AbsCudaSpinOperator())]
+
+    safe_map(ctx.replace, eqn.outvars, outvariables)
+    safe_map(ctx.write, eqn.outvars, outvals)
+
+
+def change_hamiltonian(ctx, eqn):
+    """Change catalyst hamiltonian to an equivalent expression in CUDA."""
+    assert eqn.primitive == hamiltonian_p
+
+    invals = safe_map(ctx.read, eqn.invars)
+    coeffs = invals[0]
+    terms = invals[1:]
+
+    hamiltonian = reduce(operator.add, map(operator.mul, coeffs, terms))
+
+    outvariables = [ctx.new_variable(AbsCudaSpinOperator())]
+    safe_map(ctx.replace, eqn.outvars, outvariables)
+    safe_map(ctx.write, eqn.outvars, [hamiltonian])
+
+
+def change_adjoint(ctx, eqn):
+    """Change Catalyst adjoint to an equivalent expression in CUDA."""
+    assert eqn.primitive == adjoint_p
+
+    # This is the quantum register.
+    invals = safe_map(ctx.read, eqn.invars)
+    register = invals[0]
```

Comment:
Adjoint can take arguments. At the moment this formulation only works for functions without arguments.